### ISSUE TITLE
MNIST CW attack

### ISSUE THREAD

**Issue by shairoz** [P2333/Adaptive-Diversity-Promoting#4]
Thank you for publishing your code.
According to section 4.3 and the given code, the Carlini & Wagner (C&W) attack was carried out with 1000 iterations, a learning rate of 0.01, a binary_search_steps of 1 and various confidences.
The code does not include the C&W for MNIST, when I execute a C&W on MNIST using a vanilla ResNet-56 the accuracy on adversarial examples is much higher than reported. Are these parameters used for MNIST C&W? if not can you share the C&W attack for MNIST?

Thanks

**Comment by P2333** [comment]
Thank you for the question.

The code for MNIST C&W can be obtained by slightly modifying the [data loading](https://github.com/P2333/Adaptive-Diversity-Promoting/blob/master/test_cifar_optimization.py#L27) and [input shape](https://github.com/P2333/Adaptive-Diversity-Promoting/blob/master/test_cifar_optimization.py#L63) parts in the `test_cifar_optimization.py` file. The model we used is an ensemble of three ResNet20_v1.

**Comment by shairoz** [comment]
Thank you @P2333 for your reply.
Are the C&W parameters for MNIST the same? as I've tried using the same parameters on a vanilla ResNet-56 (trained with a different code) and got significantly better accuracy

**Comment by P2333** [comment]
The parameter for MNIST can be found in Table 2, which is c=0.1; c=1; c=10. The learning rate=0.01 and iterations=1000 for CW are the same as in CIFAR-10.

**Comment by shairoz** [comment]
@P2333 I've created the following simple sanity check:
```
import keras
from keras.datasets import mnist
import cleverhans
import numpy as np
from keras.utils import to_categorical
from model import resnet_v1
from keras.layers import AveragePooling2D, Input, Flatten
import cleverhans.utils_keras
import cleverhans.attacks
import tensorflow as tf

def load_mnist():
    (X_train, y_train), (X_test, y_test) = mnist.load_data()
    X_train = np.expand_dims(X_train,-1)
    X_test = np.expand_dims(X_test,-1)
    X_train = X_train.astype(np.float32)/255.0
    X_test = X_test.astype(np.float32)/255.0
    return (X_train, y_train), (X_test, y_test)

def train_model():
    model_input = Input(shape=(28,28,1))

    model = resnet_v1(input=model_input,depth=56,num_classes=10,dataset='mnist')[0]
    (X_train, y_train), (X_test, y_test) = load_mnist()
    reduce_lr = keras.callbacks.ReduceLROnPlateau(patience=5,factor=0.1)
    checkpoint = keras.callbacks.ModelCheckpoint('./resnet_56_mnist.hdf',save_best_only=True,save_weights_only=False)
    batch_size = 128
    epochs = 60
    # optimizer = keras.optimizers.SGD(lr=0.1,momentum=0.99)
    model.compile(loss='categorical_crossentropy',
                      optimizer='adam',
                      metrics=['accuracy'])
    model.fit(X_train,to_categorical(y_train,10),
              epochs=epochs,
              batch_size=batch_size,
              callbacks=[reduce_lr,checkpoint],
              validation_data=(X_test,to_categorical(y_test,10)))

def test_cw():
    sess = tf.Session()
    keras.backend.set_session(sess)
    model = keras.models.load_model('./resnet_56_mnist.hdf')
    (X_train, y_train), (X_test, y_test) = load_mnist()
    wrap = cleverhans.utils_keras.KerasModelWrapper(model)
    attack = cleverhans.attacks.CarliniWagnerL2(wrap,sess=sess)
    adv_samples = 1000
    params = {'clip_min': 0,
              'clip_max': 1,
              'y_target': None,
              'confidence': 1,
              'batch_size': 100,
              'max_iterations': 1000,
              'binary_search_steps': 1,
              'initial_const': 0.01,
              'learning_rate': 0.01}


    p = np.argmax(model.predict(X_test),axis=1)
    ind_correct = np.where(p==y_test)[0]
    x_nor = X_test[ind_correct][:adv_samples]
    y_nor = y_test[ind_correct][:adv_samples]
    x_adv = attack.generate_np(x_nor, **params)
    adv_acc = model.evaluate(x_adv, to_categorical(y_nor, 10))[1]
    # y_adv = np.argmax(model.predict(x_adv),axis=1)
    # adv_acc = 1-(np.count_nonzero(y_nor-y_adv) / adv_samples)
    print("Adversarial ACC: %.3f" %(adv_acc))

if __name__ == '__main__':
    train_model()
    test_cw()

```
The ResNet is based on your implementation. The result is "Adversarial ACC: 1.000". Am I running it incorrectly in terms of CW parameters? The same does not occur on CIFAR10.

Thanks

**Comment by P2333** [comment]
The 'initial_const' parameter is the c in the paper. For MNIST, it should be 0.1 or 1 or 10.

In your code, the initial_const=0.01, which is too restrictive for MNIST, such that the final Adversarial ACC:1.000 (100% accuracy, i.e., no attack succeeds).

**Comment by shairoz** [comment]
Thank you @P2333 for the clarification
